packages/rest/src/validation/request-query.validator.ts

```diff
@@ -0,0 +1,150 @@
+
+  if (required && query.value == undefined) {
+    const err = Object.assign(
+      new HttpErrors.BadRequest('Request query is required'),
+      {
+        code: 'MISSING_REQUIRED_PARAMETER',
+        parameterName: 'request query',
+      },
+    );
+    throw err;
+  }
+
+  const schema = query.schema;
+  /* istanbul ignore if */
+  if (debug.enabled) {
+    debug('Request query schema: %j', util.inspect(schema, {depth: null}));
+  }
+  if (!schema) return;
+
+  options = Object.assign({coerceTypes: query.coercionRequired}, options);
+  validateValueAgainstSchema(query.value, schema, globalSchemas, options);
+}
+
+/**
+ * Convert an OpenAPI schema to the corresponding JSON schema.
+ * @param openapiSchema The OpenAPI schema to convert.
+ */
+function convertToJsonSchema(openapiSchema: SchemaObject) {
+  const jsonSchema = toJsonSchema(openapiSchema);
+  delete jsonSchema['$schema'];
+  /* istanbul ignore if */
+  if (debug.enabled) {
+    debug(
+      'Converted OpenAPI schema to JSON schema: %s',
+      util.inspect(jsonSchema, {depth: null}),
+    );
+  }
+  return jsonSchema;
+}
+
+/**
+ * Validate the request query data against JSON schema.
+ * @param query The request query data.
+ * @param schema The JSON schema used to perform the validation.
+ * @param globalSchemas Schema references.
+ */
+
+const compiledSchemaCache = new WeakMap();
+
+function validateValueAgainstSchema(
+  // tslint:disable-next-line:no-any
+  query: any,
+  schema: SchemaObject | ReferenceObject,
+  globalSchemas?: SchemasObject,
+  options?: RequestQueryValidationOptions,
+) {
+  let validate;
+
+  if (compiledSchemaCache.has(schema)) {
+    validate = compiledSchemaCache.get(schema);
+  } else {
+    validate = createValidator(schema, globalSchemas, options);
+    compiledSchemaCache.set(schema, validate);
+  }
+
+  if (validate(query)) {
+    debug('Request query passed AJV validation.');
+    return;
+  }
+
+  const validationErrors = validate.errors;
+
+  /* istanbul ignore if */
+  if (debug.enabled) {
+    debug(
+      'Invalid request query: %s. Errors: %s',
+      util.inspect(query, {depth: null}),
+      util.inspect(validationErrors),
+    );
+  }
+
+  const error = RestHttpErrors.invalidRequestQuery();
+  error.details = _.map(validationErrors, e => {
+    return {
+      path: e.dataPath,
+      code: e.keyword,
+      message: e.message,
+      info: e.params,
+    };
+  });
+  throw error;
+}
+
+function createValidator(
+  schema: SchemaObject,
+  globalSchemas?: SchemasObject,
+  options?: RequestQueryValidationOptions,
+): Function {
+  const jsonSchema = convertToJsonSchema(schema);
+
+  const schemaWithRef = Object.assign({components: {}}, jsonSchema);
+  schemaWithRef.components = {
+    schemas: globalSchemas,
+  };
+
+  const ajv = new AJV(
+    Object.assign(
+      {},
+      {
+        allErrors: true,
+      },
+      options,
+    ),
+  );
+
+  return ajv.compile(schemaWithRef);
+}
```

Comment:
There seem to be quite a bit duplicate code as request-body-validator. Can we refactor the code for better reuse?

Reply:
@raymondfeng -Thanks for the review and comments.
I will try to reply: I was actually looking to differentiate the 2 in order to allow/take into consideration future changes and needs-do you think this is un-needed? will the only difference be - the extraction of the query params and the structure to be sent to same validation function to leverage the AJV ? we need not forget, we also need different error messages.  Please share your thoughts.

Reply:
I would create a utility function that validates `ValueWithSchema` using AJV.